### PR TITLE
[RFC] [DML] INSERT, UPSERT, and REPLACE

### PR DESCRIPTION
**Pull request by am357**
*Issue #, if available:* #4

*Description of changes:*
This PR includes the RFC text for addition of PartiQL `INSERT` and `UPDATE` specification.

[rendered markdown file](https://github.com/partiql/partiql-docs/blob/3156cc48b00c8845d894ed8ae2e025cefa57b347/RFCs/0011-partiql-insert-update-spec.md)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: RFCs/11-partiql-insert-update-spec.md [hunk 1]
```diff
@@ -0,0 +1,728 @@
...
+
+This RFC specifies PartiQL's syntax and semantics for insert upsert as part of PartiQL Data Definition Language (DML).
+The RFC outlines the grammar and semantics for PartiQL’s insert or update—also known as UPSERT—and REPLACE statements.
+UPSERT is an informal term for a statement that denotes inserting or updating a set of data to a target database’s table
+based on existing availability condition of that data in the same table.
+
+# Motivation
+[motivation]: #motivation
+
+The PartiQL Specification as it stands today does not specify the syntax and semantics of Data Definition Language (DML).
```
Comment (hpolloni) [inline RFCs/11-partiql-insert-update-spec.md:16]: Data Manipulation Language?

File: RFCs/11-partiql-insert-update-spec.md [hunk 2]
```diff
@@ -0,0 +1,728 @@
...
+`SELECT` query expression can be used for inserting the resulting values from the query to the target table. The resulting tuples (e.g. rows) MUST have the same attribute layout as the target schema’s required attributes.
+
+#### 3.4 Examples
+
+The following example statement inserts items from `Vehicles` table to `HeavyVehicles` table. Both tables have the same layout for required attributes, therefore the INSERT statement can insert items using a SELECT query. The second INSERT statement shows that the attributes can be specified INSERT and SELECT.
+
+```SQL
+-- The following `CREATE TABLE` is arbitrary since the PartiQL DDL is yet to be
+-- defined.
+CREATE TABLE Music (
```
Comment (hpolloni) [inline RFCs/11-partiql-insert-update-spec.md:237]: I'm going to be the obnoxious person that ask when is DDL planned for? Some of the examples above (i.e. VALUES and DEFAULT) presume that there's a closed schema somewhere, but we don't have a facility in PartiQL to create that schema in the first place.

File: RFCs/11-partiql-insert-update-spec.md (hunk 1 above)
Comment (am357) [reply RFCs/11-partiql-insert-update-spec.md:16]: Correct, will fix.

BTW, this PR is still in draft.

File: RFCs/11-partiql-insert-update-spec.md (hunk 2 above)
Comment (am357) [reply RFCs/11-partiql-insert-update-spec.md:237]: Asking for DDL is fair, however I see that as being orthogonal to this RFC. This Spec. assumes there is a set of data-items stored in some database system that requires manipulation with PartiQL; for example, using the `INSERT` statement to manipulate data on DynamoDB. The intention for formalizing this RFC comes from a direct customer ask that's why we've prioritized. 

For DDL, as of now we haven't planned for it in short-term; would be great to have a further discussion to understand your use-case and to see how we can re-baseline our plans if needed.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 3]
```diff
@@ -0,0 +1,907 @@
+- Start Date: 2022-06-17
+- PartiQL Issue [partiql/partiql-docs/#4](https://github.com/partiql/partiql-docs/issues/4)
+- RFC PR: [partiql/partiql-docs/#11](https://github.com/partiql/partiql-docs/pull/11)
+
+# Summary
+
+[summary]: #summary
+
+This RFC specifies PartiQL's syntax and semantics for insert-or-update—also known as UPSERT—as part of PartiQL
```
Comment (amrith) [inline RFCs/0011-partiql-insert-update-spec.md:9]: Very closely related to INSERT/UPSERT is REPLACE and this PR also handles REPLACE so it would be good to call it out here.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 4]
```diff
@@ -0,0 +1,907 @@
...
+    | <collection value>
+
+<absent value ::= NULL
+    | MISSING
+
+<scalar value> :== <backtick> <ion literal> <backtick>
+    | sql literal
+
+<tuple value> ::= <l curly brace> <r curly brace>
+    | <l curly brace> <string value> : <value> [, <string value> : <value>]... <r curly brace>
```
Comment (amrith) [inline RFCs/0011-partiql-insert-update-spec.md:134]: <string value> is not defined.

File: RFCs/0011-partiql-insert-update-spec.md (hunk 3 above)
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md]: Sure, added.

File: RFCs/0011-partiql-insert-update-spec.md (hunk 4 above)
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md:136]: Discussed this offline, will iterate over the grammar.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 5]
```diff
@@ -0,0 +1,772 @@
...
+```SQL
+INSERT INTO distributors
+VALUES <<{did: 5, dname: 'Gizmo Transglobal'},
+    {did: 6, dname: 'Associated Computing, Inc'}>>
+ON CONFLICT
+    DO
+UPDATE SET dname = EXCLUDED.dname;
+```
+
+Insert distributors values as bag value. In case of a conflict on unique constraint updates the values as stated in the insert. Assumes the database engine implementation can infer the unique index as . Note that the special `excluded` table is used to reference values originally proposed for insertion:
```
Comment (dlurton) [inline RFCs/0011-partiql-insert-update-spec.md:664]: Just spotted this out of the corner of my eye:

> ... implementation can infer the unique index as .

Incomplete sentence?

File: RFCs/0011-partiql-insert-update-spec.md [hunk 6]
```diff
@@ -0,0 +1,772 @@
...
+The following example MUST lead to a `SemanticError` if there is an existing item that conflicts with the item specified by replace:
+
+```SQL
+-- Existing items is:
+--  {HK: 1, RK: 1, myAttr: 12 },
+--  {HK: 1, RK: 2, myAttr: 12 }
+-- Outcome is SemanticError
+
+INSERT into Customers
+VALUES <<{ HK: 1, RK: 1, myAttr: 12, anotherAttr: "Hello" }>>
```
Comment (dlurton) [inline RFCs/0011-partiql-insert-update-spec.md:628]: Isn't the use of `VALUES` incorrect here?

Should be:

```
INSERT into Customers <<{ HK: 1, RK: 1, myAttr: 12, anotherAttr: "Hello" }>>
```

In the `partiql-lang-kotlin` `VALUES (1, 2, 3), (4, 5, 6)` is actually syntactic alternative for `<< [1, 2, 3], [4, 5 6]>>`.

Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md:628]: Good catch, yes, it's incorrect—will fix.

File: RFCs/0011-partiql-insert-update-spec.md (hunk 5 above)
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md]: Yes, fixed in the new commit.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 7]
```diff
@@ -0,0 +1,761 @@
...
+Figure 1. illustrates the BNF grammar corresponding to PartiQL’s `INSERT INTO ... ON CONFLICT` statements. See Appendix (1) for the Grammar conventions.
+
+Note:
+
+* All the definitions that are missing from the grammar refer to the definitions that are already in SQL-92 standard.
+
+```EBNF
+<insert statement> ::= INSERT INTO <table name> [ AS <alias> ] 
+    [  ( <attr name> [, <attr name> ]... ) ]
+        <values>
```
Comment (dlurton) [inline RFCs/0011-partiql-insert-update-spec.md:60]: I *think* `<values>` should actually just be `<expr>`.  This is an extension of PartiQL/SQL++.  Note that `<expr>` covers all of `<values>` that you have defined below, and then some.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 8]
```diff
@@ -0,0 +1,761 @@
...
+
+* (a) If the number of values in a data item is equal to the number of declared attributes, the values are assumed to refer to the declared attributes, in order of their definition in the table schema, and
+
+* (b) If the number of values in data item (M) is fewer than the number of declared attributes (N) the values are assumed to refer to the first M declared attributes, in order, and
+
+* (c) If the number of values in a data item (M) is greater than the number of declared attributes (N) it is an error, and
+* (d) If a value is not specified for every required attribute, it is an error.
+
+
+It’s worth emphasizing that, the implementing database MUST provide an ordering for the declared attributes that appear in the Schema.
```
Comment (dlurton) [inline RFCs/0011-partiql-insert-update-spec.md:174]: There should be some discussion here (ish) about how the `INSERT` statement is restricted when working in a schemaless fashion.  

For example, in

```sql
INSERT INTO (id, name) VALUES (1, 'bob');
```

The list of fields preceding `VALUES` is optional according to the grammar.  But if you don't have schema, the list of fields is absolutely required.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 9]
```diff
@@ -0,0 +1,761 @@
...
+
+When attribute names are omitted and `VALUES <bag value>` is provided, the values in `<bag value>` MUST include required attribute–value pairs and MAY include optional or undeclared attribute-value pairs.
```
Comment (dlurton) [inline RFCs/0011-partiql-insert-update-spec.md:176]: There's more nuances to this, I think.  Consider

```sql
CREATE TABLE foo (
    id int not null primary key,
    is_deleted boolean not null default false,
    name varchar(50)
)
```

Then: 
```
INSERT INTO foo (1, 'bob') would not be a valid statement.
```

The statement becomes valid only if `is_deleted` is moved to the last position in the table.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 10]
```diff
@@ -0,0 +1,761 @@
...
+    'DOB' : 1974-02-10T,
+    'GovId' : '744 849 301',
+    'GovIdType' : 'SSN',
+    'Address' : '4058 Melrose Street, Spokane Valley, WA, 99206'
+    }>>;
+```
+
+### 3.2 Query
+
+`SELECT` query expression can be used for inserting the resulting values from the query to the target table. The resulting tuples (E.g. rows) MUST have the same attribute name and type as the target schema’s required attributes.
```
Comment (dlurton) [inline RFCs/0011-partiql-insert-update-spec.md:250]: PartiQL/SQL++ should be generalizing this not to just be a `select` query but *any* expression returning any PartiQL value.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 11]
```diff
@@ -0,0 +1,761 @@
...
+
+Optional attributes in target schema that have the same name in the resulting tuples from the source schema, SHOULD have the same type. If the implementing database provides type coercion rule for the types involved in source and destination, the `SELECT` MAY lead to a successful outcome, otherwise it's a `SemanticError`. See examples for more clarification.
+
+#### 3.4 Examples
+
+The following example statement inserts items from `Vehicles` table into `HeavyVehicles` table. Both tables have the same layout for required and optional attributes, therefore the `INSERT` statement can insert items using a `SELECT` query.
+
+```SQL
+-- The following `CREATE TABLE` is arbitrary since the PartiQL DDL is yet to be
+-- defined.CREATE TABLE Music
```
Comment (dlurton) [inline RFCs/0011-partiql-insert-update-spec.md:260]: newline needed.  Also why is this statement commented but not the one below?

File: RFCs/0011-partiql-insert-update-spec.md [hunk 12]
```diff
@@ -0,0 +1,761 @@
...
+INSERT INTO Music (Artist, SongTitle)
+SELECT Artist, SongTitle
+FROM RockAlbums
+WHERE RockGenre IN ('Alternative', 'SpaceRock');
+```
+
+The following example statement attempts to insert items from `Vehicles` table to `HeavyVehicles` table. Both tables have the same layout for required attributes but have different types for `Year` optional attribute, therefore—if the implementing database system does not implement a type coercion rule from `DATE` to `INT`—the `INSERT` statement using a `SELECT` query leads to a `SemanticError`.
+
+```SQL
+-- The following `CREATE TABLE` is arbitrary since the PartiQL DDL is yet to be-- defined.CREATE TABLE Music
```
Comment (dlurton) [inline RFCs/0011-partiql-insert-update-spec.md:299]: same

File: RFCs/0011-partiql-insert-update-spec.md (hunk 8 above)
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md:174]: Can you clarify what do you mean by not having schema based on the terms we've defined here? Is is an open-schema with no _declared attributes_? b/c in that case it falls into category `(c)` which is an error:
> (c) If the number of values in a data item (M) is greater than the number of declared attributes (N) it is an error, and

File: RFCs/0011-partiql-insert-update-spec.md [hunk 13]
```diff
@@ -0,0 +1,761 @@
...
+    AlbumTitle VARCHAR(25) NOT NULL,
+    Year       DATE,
+    RockGenre  VARCHAR(10),
+    PRIMARY KEY (Artist, SongTitle)
+);
+
+INSERT INTO MusicSELECT * FROM RockAlbums;
+```
+
+The following example statement, inserts some rows into table `films` from a table `tmp_films`. Both tables must have the same layout for required and optional attributes that are in common.
```
Comment (dlurton) [inline RFCs/0011-partiql-insert-update-spec.md:323]: There's some ambiguity here:  what about attributes that are *not* in common?  Is that even allowed?

File: RFCs/0011-partiql-insert-update-spec.md (hunk 9 above)
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md:176]: First off, I can see a typo here, it should be:
> When attribute names are omitted and `<bag value>` is provided ...

With that we have:
```SQL
INSERT INTO foo <<{id: 1, name: 50}>>;
```

Which intuitively should work, however based on the current spec. it must error out b/c of the following:
> ...the values in `<bag value>` MUST include required attribute–value pairs and MAY include optional or undeclared attribute-value pairs.

Having said that, I can see a nuance related to `default` value that we need to specify/call-out in this case.

Now if we want to insert using `VALUES`, it'll be:

```SQL
INSERT INTO foo VALUES (1, 'bob');
```

With this, seems like it falls into category `(b)` b/c number of values is fewer than number of _declared attributes_ (2 < 3). In that case, it'll be an `TypeMismatch` error or something like that—if the implementing NoSQL database provides the ordering from top to bottom for an Item. I don't see an issue with this though.

> (b) If the number of values in data item (M) is fewer than the number of declared attributes (N) the values are assumed to refer to the first M declared attributes, in order,

File: RFCs/0011-partiql-insert-update-spec.md (hunk 7 above)
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md:60]: I actually considered that, but intentionally ruled it out in favor of being able to assert the correctness of the grammar using an explicit definition. Happy to discuss further about this though.

File: RFCs/0011-partiql-insert-update-spec.md (hunk 9 above)
Comment (dlurton) [reply RFCs/0011-partiql-insert-update-spec.md:176]: Sorry I omitted `VALUES` from the original insert statement which was probably confusing.

File: RFCs/0011-partiql-insert-update-spec.md (hunk 10 above)
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md:250]: I think it makes sense to extend the grammar incrementally but not necessarily for all PartiQL values. For example, not sure if scalars, or `MISSING` are applicable. Having said that, to your point, I think we need a follow-up issue for this; but at this stage—for consistency—I'd rather leaving it out of scope.

File: RFCs/0011-partiql-insert-update-spec.md (hunk 13 above)
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md:324]: Yes I agree. It requires more explanation; in that case, it depends on the target schema. If it's an open schema, then yes, otherwise, no.

**Review by almann — COMMENTED** [review]
Reviewed up to L490

File: RFCs/0011-partiql-insert-update-spec.md [hunk 14]
```diff
@@ -0,0 +1,845 @@
...
+```SQL
+-- Existing items is:
+--  {HK: 1, RK: 1, myAttr: 12 },
+--  {HK: 1, RK: 2, myAttr: 12 }
+-- Outcome is SemanticError
+
+INSERT into Customers
+<<{ HK: 1, RK: 1, myAttr: 12, anotherAttr: "Hello" }>>
+ON CONFLICT
+    DO REPLACE {HK: 1, RK: 2, thirdAttr: "World"};
```
Comment (almann) [inline RFCs/0011-partiql-insert-update-spec.md:697]: Is there a valid example of `DO REPLACE EXCLUDED`?

**Review by almann — CHANGES_REQUESTED** [review]
Overall this is looks reasonable.  A few top-line things:

* I think we need to clarify what specifying columns (basically DML schema overlay) means for `VALUES ....` expressions and their equivalent `<<[...], [...], ...>>` syntax.
* We also should be crisp as to what `DO UPDATE SET <tuple-value>` or even if that should be allowed.  The verb `SET` for this is very confusing to me (at the risk of bike shedding, `DO REPLACE VALUE <tuple-value>` makes more sense than using `SET`).
* I didn't see examples for `DO REPLACE SET <tuple-value>` which is the form that makes more sense over `DO UPDATE SET <tuple-value>`.

File: RFCs/0011-partiql-insert-update-spec.md (hunk 14 above)
Comment (almann) [reply RFCs/0011-partiql-insert-update-spec.md:697]: Also doesn't the grammar dictate that it is `DO REPLACE SET {HK: 1, RK: 2, thirdAttr: "World"}`?

File: RFCs/0011-partiql-insert-update-spec.md [hunk 15]
```diff
@@ -0,0 +1,845 @@
...
+--  {HK: 1, RK: 1, myAttr: 10}
+-- Item after the update:
+--  {HK: 1, RK: 1, myAttr: 12, anotherAttr: "Hello"}
+
+INSERT into Customers
+<<
+{HK: 1, RK: 1, myAttr: 12, anotherAttr: "Hello"}
+>>
+ON CONFLICT DO
+UPDATE EXCLUDED
```
[the remark this answers is not in the document]
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md:542]: Correct—this is to reduce the churn for duplicating the values in the `ON CONFLICT`.

**Review by almann — COMMENTED** [review]
Looks pretty good--I do have a concern with `DEFAULT` that we should decide on.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 16]
```diff
@@ -252,74 +305,64 @@ INSERT INTO Person
...
+'FirstName' : NULL,
+'DOB' : 1967-07-03T,
+'Address' : '43 Stockert Hollow Road, Everett, WA, 98203'
+},{'LastName' : 'Pena',
+'FirstName' : NULL,
+'DOB' : 1974-02-10T,
+'GovId' : '744 849 301',
+'GovIdType' : 'SSN',
+'Address' : '4058 Melrose Street, Spokane Valley, WA, 99206'
+}
```
[the remark this answers is not in the document]
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md:322]: Makes sense—changed.

File: RFCs/0011-partiql-insert-update-spec.md (hunk 11 above)
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md]: This is fixed in new commits.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 17]
```diff
@@ -340,26 +384,28 @@ CREATE TABLE RockAlbums
...
 ```
 
-The following statement shows that the values can be specified with a sub-select.
+The following statement shows that the values can be specified with a sub-select. The INSERT goes through because
+the values that are not `SELECT` default to `NULL`.
 
 ```SQL
 INSERT INTO Music (Artist, SongTitle)
-SELECT Artist, SongTitle
+SELECT Artist, SongTitle, AlbumTitle
```
[the remark this answers is not in the document]
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md]: Yes, missed it. Fixed.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 18]
```diff
@@ -384,14 +431,78 @@ INSERT INTO Music
...
+   is_deleted BOOLEAN NOT NULL DEFAULT FALSE,
+   title      VARCHAR(50),
+   bar        VARCHAR(10) DEFAULT "baz"
+);
+
+INSERT INTO Foo (id, title)
+<<
+[1, DEFAULT],
+[2, "some-name"],
+>>;
```
[the remark this answers is not in the document]
Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md]: I think this diff is pointing to a commit that is related to `DO REPLACE` changes; this has been removed from the final diff.

**Review by almann — DISMISSED** [review]
Thanks for the hard work on this--looks good.

**Review by amrith — APPROVED** [review]
Looks good Arash, thanks for patiently working through the comments.

**Review by amrith — APPROVED** [review]
LGTM -- oops, duplicate.

File: RFCs/0011-partiql-insert-update-spec.md [hunk 19]
```diff
@@ -185,9 +185,11 @@ Optional attributes in target schema that have the same name in the resulting tu
 
 In the following examples, `Films` table has a closed schema. In the table `len` attribute is implicitly defined as nullable.
 
+_In the following examples `CREATE TABLE` including `SCHEMA OPEN` and `SCHEMA CLOSED` syntax is arbitrary since the PartiQL DDL is yet to be defined._
+
```
Comment (almann) [inline RFCs/0011-partiql-insert-update-spec.md:189]: *Minor*: I would define what these terms mean.  Here is what I might write for this:

`SCHEMA OPEN` means that the struct/tuple may have attributes not defined in the schema and allows for defined attributes to be _optional_. `SCHEMA CLOSED` means that the attributes defined must all be present and there is no _optional_ attributes (i.e. all attributes MUST have a value, even if that value is `NULL`).

**Review by almann — DISMISSED** [review]
Looks good--minor edit suggested below, but not required if you think it too wordy.

**Review by almann — APPROVED** [review]
:shipit:

File: RFCs/0011-partiql-insert-update-spec.md [hunk 20]
```diff
@@ -0,0 +1,1041 @@
...
+Figure 1. illustrates the BNF grammar corresponding to PartiQL’s `INSERT INTO ... ON CONFLICT` statements. See Appendix (1) for the Grammar conventions.
+
+Note:
+
+* All the definitions that are missing from the grammar refer to the definitions that are already in SQL-92 standard.
+
+```EBNF
+<insert statement> ::= INSERT INTO <table name> [ AS <alias> ] 
+    [  ( <attr name> [, <attr name> ]... ) ]
+        <values>
```
Comment (johnedquinn) [inline RFCs/0011-partiql-insert-update-spec.md:56]: Leaving a post-merge comment here: Since we use `EXCLUDED` as a variable name that references a value to be inserted, we could alternatively propose a follow-up RFC that specifies the syntax to allow an alias for the value to be inserted. Such as:
```ebnf
<insert statement> ::= INSERT INTO <table name> [ AS <alias> ] 
    [  ( <attr name> [, <attr name> ]... ) ]
        <values> [ AS <alias> ]
    ...
```

Note the `[ AS <alias> ]` after `<values>`

Comment (am357) [reply RFCs/0011-partiql-insert-update-spec.md:56]: @johnedquinn it's an intriguing idea: what do you consider the value it brings in addition to `EXCLUDED`?
